src/background/restrictUnauthenticatedUrlAccess.ts

```diff
@@ -42,19 +44,55 @@ async function getAuthUrlPatterns(organizationId: UUID): Promise<string[]> {
   }
 }
 
+async function isRestrictedUrl(url: string) {
+  const cachedAuthUrlPatterns = await sessionValue.get();
+  return testMatchPatterns(cachedAuthUrlPatterns, url);
+}
+
 function getDefaultAuthUrl(restrictedUrl: string) {
   const errorMessage = `Access is restricted to '${restrictedUrl}'. Log in with PixieBrix to proceed`;
   const defaultUrl = new URL("https://app.pixiebrix.com/login/");
   defaultUrl.searchParams.set("error", errorMessage);
   return defaultUrl.href;
 }
 
+async function getRedirectUrl(restrictedUrl: string) {
+  const { ssoUrl } = await readManagedStorage();
+  return ssoUrl ?? getDefaultAuthUrl(restrictedUrl);
+}
+
+async function redirectRestrictedTab({
+  tabId,
+  url,
+}: {
+  tabId: number;
+  url: string;
+}) {
+  if (await isRestrictedUrl(url)) {
+    await browser.tabs.update(tabId, {
+      url: await getRedirectUrl(url),
+    });
+  }
+}
+
+async function handleRestrictedTab(
+  tabId: number,
+  changeInfo: Tabs.OnUpdatedChangeInfoType,
+  tab: Tabs.Tab,
+) {
+  if (!changeInfo.url || !tab?.url) {
```

Comment:
tab.url can't be falsy if changeInfo.url is defined 
```suggestion
  if (!changeInfo.url) {
```

Reply:
Really? It's impossible for tab.url to get unset?

Reply:
I misread your comment the first time, this makes sense